File: research/src/features.py
"""
Feature computation — pure functions over BookState + TradeWindow + MidHistory.

Every function matches the spec in research/spec/features.json. The TS inference
adapter (src/research/inference/features.ts) implements the same formulas — the
parity test verifies they produce identical output.
"""

import math

from .data import BookState, TradeWindow, MidHistory


# ── Microstructure (book-only) ───────────────────────────────────────────────


def spread_bps(book: BookState) -> float:
    mid = book.mid()
    if mid is None or mid == 0:
        return 0.0
    return (book.best_ask - book.best_bid) / mid * 10_000


def microprice(book: BookState) -> float:
    tb = book.top_bids(1)
    ta = book.top_asks(1)
    if not tb or not ta:
        return 0.0
    bid_p, bid_s = tb[0]
    ask_p, ask_s = ta[0]
    total = bid_s + ask_s
    if total == 0:
        return 0.0
    return (bid_p * ask_s + ask_p * bid_s) / total


def book_imbalance(book: BookState, levels: int = 1) -> float:
    bids = book.top_bids(levels)
    asks = book.top_asks(levels)
    bid_qty = sum(s for _, s in bids)
    ask_qty = sum(s for _, s in asks)
    total = bid_qty + ask_qty
    if total == 0:
        return 0.0
    return (bid_qty - ask_qty) / total


def depth_bid(book: BookState, levels: int = 5) -> float:
    return sum(s for _, s in book.top_bids(levels))


def depth_ask(book: BookState, levels: int = 5) -> float:
    return sum(s for _, s in book.top_asks(levels))


def depth_ratio(book: BookState, levels: int = 5) -> float:
    d_ask = depth_ask(book, levels)
    if d_ask == 0:
        return 0.0
    return depth_bid(book, levels) / d_ask


def wap_distance_bps(book: BookState, levels: int = 5) -> float:
    """Effective spread of top-N depth measured as VWAP distance in bps."""
    mid = book.mid()
    if mid is None or mid == 0:
        return 0.0

    bids = book.top_bids(levels)
    asks = book.top_asks(levels)

    bid_notional = sum(p * s for p, s in bids)
    bid_qty = sum(s for _, s in bids)
    ask_notional = sum(p * s for p, s in asks)
    ask_qty = sum(s for _, s in asks)

    if bid_qty == 0 or ask_qty == 0:
        return 0.0

    vwap_bid = bid_notional / bid_qty
    vwap_ask = ask_notional / ask_qty
    return (vwap_ask - vwap_bid) / mid * 10_000


# ── Trade flow (requires TradeWindow) ────────────────────────────────────────


def trade_imbalance(tw: TradeWindow, now: int, window_ms: int = 60_000) -> float:
    trades = tw.trades_in_window(now, window_ms)
    if not trades:
        return 0.0
    buy_vol = sum(t.size for t in trades if t.side == "bid")
    sell_vol = sum(t.size for t in trades if t.side == "ask")
    total = buy_vol + sell_vol
    if total == 0:
        return 0.0
    return (buy_vol - sell_vol) / total


def trade_intensity(tw: TradeWindow, now: int, window_ms: int = 60_000) -> float:
    trades = tw.trades_in_window(now, window_ms)
    if window_ms == 0:
        return 0.0
    return len(trades) / (window_ms / 1000)


def avg_trade_size(tw: TradeWindow, now: int, window_ms: int = 60_000) -> float:
    trades = tw.trades_in_window(now, window_ms)
    if not trades:
        return 0.0
    return sum(t.size for t in trades) / len(trades)


def ofi(tw: TradeWindow, now: int, window_ms: int = 60_000) -> float:
    """
    Order flow imbalance: net signed trade volume.

    Positive = net buying pressure; negative = net selling pressure.
    Normalized by total volume to be scale-independent.
    """
    trades = tw.trades_in_window(now, window_ms)
    if not trades:
        return 0.0
    signed = sum(t.size if t.side == "bid" else -t.size for t in trades)
    total = sum(t.size for t in trades)
    if total == 0:
        return 0.0
    return signed / total


# ── Volatility (requires MidHistory) ─────────────────────────────────────────


def realized_vol(mh: MidHistory, now: int, window_ms: int = 300_000) -> float:
    """Realized volatility from mid returns, annualized."""
    entries = mh.mids_in_window(now, window_ms)
    if len(entries) < 2:
        return 0.0
    log_returns = []
    for i in range(1, len(entries)):
        if entries[i - 1][1] > 0 and entries[i][1] > 0:
            log_returns.append(math.log(entries[i][1] / entries[i - 1][1]))
    if len(log_returns) < 2:
        return 0.0
    mean = sum(log_returns) / len(log_returns)
    var = sum((r - mean) ** 2 for r in log_returns) / (len(log_returns) - 1)
    # Annualize: assume entries are ~1s apart, scale by sqrt(seconds_per_year)
    dt_avg = (entries[-1][0] - entries[0][0]) / max(len(entries) - 1, 1) / 1000
    if dt_avg <= 0:
        return 0.0
    periods_per_year = 365.25 * 24 * 3600 / dt_avg
    return math.sqrt(var * periods_per_year)


def log_return(mh: MidHistory, now: int, window_ms: int = 10_000) -> float:
    """Log return of mid over window."""
    current = mh.mid_at_or_before(now)
    past = mh.mid_at_or_before(now - window_ms)
    if current is None or past is None or past <= 0 or current <= 0:
        return 0.0
    return math.log(current / past)


# ── Compute all features ─────────────────────────────────────────────────────


def compute_all(
    book: BookState,
    tw: TradeWindow,
    mh: MidHistory,
    now: int,
) -> dict[str, float]:
    """Compute the full feature vector. Keys match research/spec/features.json names."""
    return {
        "spread_bps": spread_bps(book),
        "microprice": microprice(book),
        "book_imbalance_1": book_imbalance(book, levels=1),
        "book_imbalance_5": book_imbalance(book, levels=5),
        "depth_bid_5": depth_bid(book, levels=5),
        "depth_ask_5": depth_ask(book, levels=5),
        "depth_ratio_5": depth_ratio(book, levels=5),
        "wap_distance_bps": wap_distance_bps(book, levels=5),
        "trade_imbalance_60s": trade_imbalance(tw, now, 60_000),
        "trade_intensity_60s": trade_intensity(tw, now, 60_000),
        "avg_trade_size_60s": avg_trade_size(tw, now, 60_000),
        "ofi_60s": ofi(tw, now, 60_000),
        "realized_vol_300s": realized_vol(mh, now, 300_000),
        "return_10s": log_return(mh, now, 10_000),
        "return_60s": log_return(mh, now, 60_000),
    }


FEATURE_NAMES = list(compute_all.__code__.co_consts)  # not used at runtime
FEATURE_NAMES = [
    "spread_bps", "microprice", "book_imbalance_1", "book_imbalance_5",
    "depth_bid_5", "depth_ask_5", "depth_ratio_5", "wap_distance_bps",
    "trade_imbalance_60s", "trade_intensity_60s", "avg_trade_size_60s",
    "ofi_60s", "realized_vol_300s", "return_10s", "return_60s",
]

File: src/research/inference/features.ts
/**
 * Feature computer — computes the same features as research/src/features.py
 * from a LocalBook + rolling trade/mid buffers.
 *
 * The parity test (research/tests/test_parity.py) verifies these produce
 * identical output to the Python side on the same input.
 */
import type { LocalBook } from "../../data/feed.js";

export interface TradeRecord {
  ts: number;
  side: string;
  price: number;
  size: number;
}

export interface FeatureVector {
  spread_bps: number;
  microprice: number;
  book_imbalance_1: number;
  book_imbalance_5: number;
  depth_bid_5: number;
  depth_ask_5: number;
  depth_ratio_5: number;
  wap_distance_bps: number;
  trade_imbalance_60s: number;
  trade_intensity_60s: number;
  avg_trade_size_60s: number;
  ofi_60s: number;
  realized_vol_300s: number;
  return_10s: number;
  return_60s: number;
}

export const FEATURE_NAMES: (keyof FeatureVector)[] = [
  "spread_bps", "microprice", "book_imbalance_1", "book_imbalance_5",
  "depth_bid_5", "depth_ask_5", "depth_ratio_5", "wap_distance_bps",
  "trade_imbalance_60s", "trade_intensity_60s", "avg_trade_size_60s",
  "ofi_60s", "realized_vol_300s", "return_10s", "return_60s",
];

// ── Book features ───────────────────────────────────────────────────────────

function topBids(book: LocalBook, n: number): [number, number][] {
  return [...book.bids.entries()].sort((a, b) => b[0] - a[0]).slice(0, n);
}

function topAsks(book: LocalBook, n: number): [number, number][] {
  return [...book.asks.entries()].sort((a, b) => a[0] - b[0]).slice(0, n);
}

function mid(book: LocalBook): number | null {
  if (book.bestBid === -Infinity || book.bestAsk === Infinity) return null;
  return (book.bestBid + book.bestAsk) / 2;
}

function spreadBps(book: LocalBook): number {
  const m = mid(book);
  if (m === null || m === 0) return 0;
  return (book.bestAsk - book.bestBid) / m * 10_000;
}

function micropriceCalc(book: LocalBook): number {
  const tb = topBids(book, 1);
  const ta = topAsks(book, 1);
  if (tb.length === 0 || ta.length === 0) return 0;
  const [bidP, bidS] = tb[0];
  const [askP, askS] = ta[0];
  const total = bidS + askS;
  if (total === 0) return 0;
  return (bidP * askS + askP * bidS) / total;
}

function bookImbalance(book: LocalBook, levels: number): number {
  const bids = topBids(book, levels);
  const asks = topAsks(book, levels);
  let bidQty = 0;
  let askQty = 0;
  for (const [, s] of bids) bidQty += s;
  for (const [, s] of asks) askQty += s;
  const total = bidQty + askQty;
  if (total === 0) return 0;
  return (bidQty - askQty) / total;
}

function depthBid(book: LocalBook, levels: number): number {
  let sum = 0;
  for (const [, s] of topBids(book, levels)) sum += s;
  return sum;
}

function depthAsk(book: LocalBook, levels: number): number {
  let sum = 0;
  for (const [, s] of topAsks(book, levels)) sum += s;
  return sum;
}

function depthRatio(book: LocalBook, levels: number): number {
  const da = depthAsk(book, levels);
  if (da === 0) return 0;
  return depthBid(book, levels) / da;
}

function wapDistanceBps(book: LocalBook, levels: number): number {
  const m = mid(book);
  if (m === null || m === 0) return 0;

  const bids = topBids(book, levels);
  const asks = topAsks(book, levels);

  let bidNotional = 0, bidQty = 0;
  for (const [p, s] of bids) { bidNotional += p * s; bidQty += s; }
  let askNotional = 0, askQty = 0;
  for (const [p, s] of asks) { askNotional += p * s; askQty += s; }

  if (bidQty === 0 || askQty === 0) return 0;
  const vwapBid = bidNotional / bidQty;
  const vwapAsk = askNotional / askQty;
  return (vwapAsk - vwapBid) / m * 10_000;
}

// ── Trade flow features ─────────────────────────────────────────────────────

function tradesInWindow(trades: TradeRecord[], now: number, windowMs: number): TradeRecord[] {
  const cutoff = now - windowMs;
  return trades.filter(t => t.ts >= cutoff);
}

function tradeImbalance(trades: TradeRecord[], now: number, windowMs: number): number {
  const w = tradesInWindow(trades, now, windowMs);
  if (w.length === 0) return 0;
  let buy = 0, sell = 0;
  for (const t of w) {
    if (t.side === "bid") buy += t.size;
    else sell += t.size;
  }
  const total = buy + sell;
  if (total === 0) return 0;
  return (buy - sell) / total;
}

function tradeIntensity(trades: TradeRecord[], now: number, windowMs: number): number {
  if (windowMs === 0) return 0;
  const w = tradesInWindow(trades, now, windowMs);
  return w.length / (windowMs / 1000);
}

function avgTradeSize(trades: TradeRecord[], now: number, windowMs: number): number {
  const w = tradesInWindow(trades, now, windowMs);
  if (w.length === 0) return 0;
  let sum = 0;
  for (const t of w) sum += t.size;
  return sum / w.length;
}

function ofiCalc(trades: TradeRecord[], now: number, windowMs: number): number {
  const w = tradesInWindow(trades, now, windowMs);
  if (w.length === 0) return 0;
  let signed = 0, total = 0;
  for (const t of w) {
    signed += t.side === "bid" ? t.size : -t.size;
    total += t.size;
  }
  if (total === 0) return 0;
  return signed / total;
}

// ── Volatility features ─────────────────────────────────────────────────────

interface MidEntry { ts: number; mid: number; }

function midsInWindow(mids: MidEntry[], now: number, windowMs: number): MidEntry[] {
  const cutoff = now - windowMs;
  return mids.filter(e => e.ts >= cutoff);
}

function midAtOrBefore(mids: MidEntry[], targetTs: number): number | null {
  let result: number | null = null;
  for (const e of mids) {
    if (e.ts <= targetTs) result = e.mid;
    else break;
  }
  return result;
}

function realizedVol(mids: MidEntry[], now: number, windowMs: number): number {
  const entries = midsInWindow(mids, now, windowMs);
  if (entries.length < 2) return 0;
  const logReturns: number[] = [];
  for (let i = 1; i < entries.length; i++) {
    if (entries[i - 1].mid > 0 && entries[i].mid > 0) {
      logReturns.push(Math.log(entries[i].mid / entries[i - 1].mid));
    }
  }
  if (logReturns.length < 2) return 0;
  const mean = logReturns.reduce((s, r) => s + r, 0) / logReturns.length;
  const variance = logReturns.reduce((s, r) => s + (r - mean) ** 2, 0) / (logReturns.length - 1);
  const dtAvg = (entries[entries.length - 1].ts - entries[0].ts) / Math.max(entries.length - 1, 1) / 1000;
  if (dtAvg <= 0) return 0;
  const periodsPerYear = 365.25 * 24 * 3600 / dtAvg;
  return Math.sqrt(variance * periodsPerYear);
}

function logReturn(mids: MidEntry[], now: number, windowMs: number): number {
  const current = midAtOrBefore(mids, now);
  const past = midAtOrBefore(mids, now - windowMs);
  if (current === null || past === null || past <= 0 || current <= 0) return 0;
  return Math.log(current / past);
}

// ── Rolling state manager ───────────────────────────────────────────────────

export class FeatureState {
  private trades: TradeRecord[] = [];
  private mids: MidEntry[] = [];
  private readonly maxWindowMs: number;

  constructor(maxWindowMs = 300_000) {
    this.maxWindowMs = maxWindowMs;
  }

  addTrade(t: TradeRecord): void {
    this.trades.push(t);
  }

  addMid(ts: number, midVal: number): void {
    this.mids.push({ ts, mid: midVal });
  }

  prune(now: number): void {
    const cutoff = now - this.maxWindowMs;
    while (this.trades.length > 0 && this.trades[0].ts < cutoff) this.trades.shift();
    while (this.mids.length > 0 && this.mids[0].ts < cutoff) this.mids.shift();
  }

  compute(book: LocalBook, now: number): FeatureVector {
    return {
      spread_bps: spreadBps(book),
      microprice: micropriceCalc(book),
      book_imbalance_1: bookImbalance(book, 1),
      book_imbalance_5: bookImbalance(book, 5),
      depth_bid_5: depthBid(book, 5),
      depth_ask_5: depthAsk(book, 5),
      depth_ratio_5: depthRatio(book, 5),
      wap_distance_bps: wapDistanceBps(book, 5),
      trade_imbalance_60s: tradeImbalance(this.trades, now, 60_000),
      trade_intensity_60s: tradeIntensity(this.trades, now, 60_000),
      avg_trade_size_60s: avgTradeSize(this.trades, now, 60_000),
      ofi_60s: ofiCalc(this.trades, now, 60_000),
      realized_vol_300s: realizedVol(this.mids, now, 300_000),
      return_10s: logReturn(this.mids, now, 10_000),
      return_60s: logReturn(this.mids, now, 60_000),
    };
  }

  toArray(fv: FeatureVector): number[] {
    return FEATURE_NAMES.map(k => fv[k]);
  }
}
